Homework 4, 5 and 6

GibranAlvarez opened the pull request:
## What changes are you trying to make? (e.g. Adding or removing code, refactoring existing code, adding reports)
Creating a logical for Homework 1 and SQL queries for Homework 2
## What did you learn from the changes you have made?
to create logical models and to code queries
## Was there another approach you were thinking about making? If so, what approach(es) were you thinking of?
not for the moment but with more practice I think will be able
## Were there any challenges? If so, what issue(s) did you face? How did you overcome it?
with some syntax on the queries but looking on google and trial error I managed to fix them
## How were these changes tested?
for the queries, checking that they work and that the generated results where in fact those that the homework required
## A reference to a related issue in your repository (if applicable)

## Checklist
- [ ] I can confirm that my changes are working as intended

GibranAlvarez committed:
Homework 1.drawio

GibranAlvarez committed:
Added Homework 1.drawio

GibranAlvarez committed:
Delete 03_homework/Untitled Diagram.drawio

wrong file

GibranAlvarez committed:
Uploaded Homework 2

GibranAlvarez committed:
Updated Homework 2

amanda-ng518 reviewed (changes requested):
Hi there, I can't view your HW1 picture or pdf

GibranAlvarez commented:
Hi Amanda; is the Homework 1.drawio  file, sorry I saved it directly from the website since it had an option here is a screenshot
![image](https://github.com/GibranAlvarez/sql/assets/165569283/7a0404c9-523c-49c2-b01b-19bd290c097b)

amanda-ng518 reviewed (approved):
HW1: (reviewed screenshot version of it), nice diagram, 8/8

amanda-ng518 reviewed (approved):
HW2:
good job
8/8

GibranAlvarez committed:
Added Homework 3

amanda-ng518 reviewed (approved):
HW3: good
AGGREGATE Q2 may use LEFT JOIN
8/8

GibranAlvarez committed:
Added Homework 4

GibranAlvarez committed:
Added Homework 4

GibranAlvarez committed:
Added Homework 5

GibranAlvarez committed:
Added homework_6.md

GibranAlvarez committed:
Updated Homework 5

GibranAlvarez committed:
commit
Merge branch 'Homework-1' of https://github.com/GibranAlvarez/sql into Homework-1

GibranAlvarez committed:
Updated orthography of homework_6.

amanda-ng518 reviewed (approved):
HW4: 8/8

amanda-ng518 reviewed (approved):
HW6: 8/8

amanda-ng518 reviewed (approved):
HW5: 8/8
INSERT Q2
INSERT INTO product_units (product_id, product_name,
  product_size, product_category_id, product_qty_type, snapshot_timestamp)
      SELECT
          product_id,
          product_name,
          product_size,
          product_category_id,
          product_qty_type,
          CURRENT_TIMESTAMP
      FROM product AS p
      WHERE product_id = 7

DELETE Q1
WITH older_record AS
(SELECT product_id
		,MIN(snapshot_timestamp) AS snapshot_timestamp
		FROM product_units 
		WHERE product_id = 7)
DELETE FROM product_units
WHERE (product_id = 
	(SELECT  product_id 
	FROM older_record
		)
AND snapshot_timestamp = 	
	(SELECT  snapshot_timestamp 
	FROM older_record
		))

UPDATE Q1
ALTER TABLE product_units
ADD current_quantity INT;
UPDATE product_units
SET current_quantity = (
	SELECT current_quantity 
	FROM (
		SELECT  p.product_id, COALESCE(quantity,0) as current_quantity
		FROM product_units p
		LEFT JOIN (
			SELECT *
			,ROW_NUMBER() OVER( PARTITION BY vi.product_id ORDER BY market_date DESC) AS rn
			FROM vendor_inventory vi 
		) vi ON p.product_id  = vi.product_id
		WHERE rn = 1 
		OR rn IS NULL
	) p
WHERE product_units.product_id = p.product_id);
